src/Data/Formatter/DateTime.purs

```diff
@@ -0,0 +1,426 @@
+module Data.Formatter.DateTime
+  ( Formatter
+  , FormatterF(..)
+  , printFormatter
+  , parseFormatString
+  , format
+  , formatDateTime
+  , unformat
+  , unformatDateTime
+  ) where
+
+import Prelude
+
+import Control.Bind ((=<<))
+import Control.Lazy as Lazy
+import Control.Monad.State (State, runState, put, modify)
+import Control.Monad.Trans (lift)
+
+import Data.Int as Int
+import Data.Bifunctor (lmap)
+import Data.Functor (($>))
+import Data.Functor.Mu (Mu, unroll, roll)
+import Data.Either (Either(..))
+import Data.String as Str
+import Data.Array (some)
+import Data.Array as Arr
+import Data.DateTime as DT
+import Data.Date as D
+import Data.Time as T
+import Data.DateTime.Instant (instant, toDateTime, fromDateTime, unInstant)
+import Data.Time.Duration as Dur
+import Data.Enum (fromEnum, toEnum)
+import Data.Tuple (Tuple(..))
+import Data.Maybe (Maybe(..), maybe, isJust, fromMaybe)
+
+import Text.Parsing.Parser as P
+import Text.Parsing.Parser.Combinators as PC
+import Text.Parsing.Parser.String as PS
+import Text.Parsing.Parser.Pos (initialPos)
+
+import Utils (digit, foldDigits)
+
+data FormatterF a
+  = YearFull a
+  | YearTwoDigits a
+  | YearAbsolute a
+  | MonthFull a
+  | MonthShort a
+  | MonthTwoDigits a
+  | DayOfMonth a
+  | UnixTimestamp a
+  | DayOfWeek a
+  | Hours24 a
+  | Hours12 a
+  | Meridiem a
+  | Minutes a
+  | Seconds a
+  | Milliseconds a
+  | Placeholder String a
+  | End
+
+instance formatterFFunctor ∷ Functor FormatterF where
+  map f (YearFull a) = YearFull $ f a
+  map f (YearTwoDigits a) = YearTwoDigits $ f a
+  map f (YearAbsolute a) = YearAbsolute $ f a
+  map f (MonthFull a) = MonthFull $ f a
+  map f (MonthShort a) = MonthShort $ f a
+  map f (MonthTwoDigits a) = MonthTwoDigits $ f a
+  map f (DayOfMonth a) = DayOfMonth $ f a
+  map f (UnixTimestamp a) = UnixTimestamp $ f a
+  map f (DayOfWeek a) = DayOfWeek $ f a
+  map f (Hours24 a) = Hours24 $ f a
+  map f (Hours12 a) = Hours12 $ f a
+  map f (Meridiem a) = Meridiem $ f a
+  map f (Minutes a) = Minutes $ f a
+  map f (Seconds a) = Seconds $ f a
+  map f (Milliseconds a) = Milliseconds $ f a
+  map f (Placeholder str a) = Placeholder str $ f a
+  map f End = End
+
+type Formatter = Mu FormatterF

```

Comment:
What's the idea with using `Mu` here, rather than something like `newtype Formatter = List FormatterF`? I mean, it definitely works! But it wouldn't have occurred to me to use `Mu` for any non-tree structure.